src/py/flwr/common/serde_test.py

```diff
@@ -387,3 +419,142 @@ def test_message_to_and_from_taskres(
     if original.has_error():
         assert original.error == deserialized.error
     assert metadata == deserialized.metadata
+
+
+def test_metadata_serialization_deserialization() -> None:
+    """Test serialization and deserialization of Metadata."""
+    # Prepare
+    maker = RecordMaker()
+    original = maker.metadata()
+
+    # Execute
+    proto = metadata_to_proto(original)
+    deserialized = metadata_from_proto(proto)
+
+    # Assert
+    assert isinstance(proto, ProtoMetadata)
+    assert original == deserialized
+
+
+@pytest.mark.parametrize(
+    "content_fn, error_fn",
+    [
+        (
+            lambda maker: maker.recordset(1, 1, 1),
+            None,
+        ),  # check when only content is set
+        (None, lambda code: Error(code=code)),  # check when only error is set
+    ],
+)
+def test_message_serialization_deserialization(
+    content_fn: Callable[
+        [
+            RecordMaker,
+        ],
+        RecordSet,
+    ],
+    error_fn: Callable[[int], Error],
+) -> None:
+    """Test serialization and deserialization of Message."""
+    # Prepare
+    maker = RecordMaker(state=2)
+    metadata = maker.metadata()
+    metadata.dst_node_id = 0  # Assume driver node
+
+    original = Message(
+        metadata=metadata,
+        content=None if content_fn is None else content_fn(maker),
+        error=None if error_fn is None else error_fn(0),
+    )
+
+    # Execute
+    proto = message_to_proto(original)
+    deserialized = message_from_proto(proto)
+
+    # Assert
+    assert isinstance(proto, ProtoMessage)
+
+    if original.has_content():
+        assert original.content == deserialized.content
+    if original.has_error():
+        assert original.error == deserialized.error
+
+    # Deserialized metadata happens after original metadata creation
+    assert deserialized.metadata.created_at > metadata.created_at
```

Comment:
This should be the same. It seems we need to track metadata's `created_at` as part of its proto representation. I did this in this PR: https://github.com/adap/flower/pull/4012